src/common/snippets/src/lowered/pass/solve_buffer_memory.cpp

```diff
@@ -14,73 +15,144 @@ namespace snippets {
 namespace lowered {
 namespace pass {
 
-std::vector<ov::MemorySolver::Box> SolveBufferMemory::init_boxes(const AllocateBuffers::BufferClusters& buffer_clusters) {
-    std::vector<ov::MemorySolver::Box> boxes;
-    const auto count = static_cast<int>(buffer_clusters.size());
-    for (int i = 0; i < count; i++) {
-        ov::MemorySolver::Box box = { std::numeric_limits<int>::max(), 0, 0, i };
-        int64_t box_size = 0;
-        for (const auto& buffer_expr : buffer_clusters[i]) {
-            int e_start = 0, e_finish = 0;
-            const auto buffer = ov::as_type_ptr<ov::snippets::op::Buffer>(buffer_expr->get_node());
-            OPENVINO_ASSERT(buffer != nullptr, "BufferSolver expects Buffer ops in clusters");
-
-            // life finish time - order of LoopEnd / MemoryAccess ops
-            const auto& buffer_outs = buffer_expr->get_output_port_connectors();
-            for (const auto& buffer_out : buffer_outs) {
-                const auto consumers = buffer_out->get_consumers();
-                for (const auto& consumer : consumers) {
-                    const auto consumer_order = static_cast<int>(ov::snippets::pass::GetTopologicalOrder(consumer.get_expr()->get_node()));
-                    e_finish = std::max(e_finish, consumer_order);  // the last consumer
-                }
+std::pair<LinearIR::container, LinearIR::container> SolveBufferMemory::extract_static_and_dynamic_buffers(const LinearIR::container& buffer_expressions) {
+    LinearIR::container static_buffer_exprs, dynamic_buffer_exprs;
+    for (const auto& buffer_expr : buffer_expressions) {
+        const auto& buffer = ov::as_type_ptr<op::Buffer>(buffer_expr->get_node());
+        OPENVINO_ASSERT(buffer, "Buffer clusters expects Buffer nodes");
+
+        auto& clusters = buffer->is_defined() ? static_buffer_exprs : dynamic_buffer_exprs;
+        clusters.push_back(buffer_expr);
+    }
+
+    // Validation check that buffer cluster has only static or dynamic buffers.
+    for (const auto& static_buffer : static_buffer_exprs) {
+        const auto static_cluster_id = ov::as_type_ptr<op::Buffer>(static_buffer->get_node())->get_cluster_id();
+        auto is_cluster_ids_the_same = [&static_cluster_id](const ExpressionPtr& expr) {
+            return static_cluster_id == ov::as_type_ptr<op::Buffer>(expr->get_node())->get_cluster_id();
+        };
+        OPENVINO_ASSERT(std::none_of(dynamic_buffer_exprs.cbegin(), dynamic_buffer_exprs.cend(), is_cluster_ids_the_same),
+                        "There is Buffer cluster with buffers which has defined and undefined allocation sizes");
+    }
```

Comment:
We have a bunch of validation passes, and it seems that these kinds of checks should be performed there.